File: Data Structures & Algorithms/string-encode-and-decode/submission-3.cpp
class Solution {
public:

    string encode(vector<string>& strs) {
        string ans = "";
        int x = strs.size();
        string s = "";
        for(int i=0; i<x; i++){
            s += to_string(strs[i].size()) + ',';
        }
        s += '#';
        for(auto ele: strs) s += ele;
        cout<<s<<endl;
        return s;
    }

    vector<string> decode(string s) {
        vector<string> ans;
        int x = s.length();
        vector<int> len;
        string si = "";
        int brk = 0;
        for(int i=0; i<x; i++){
            if(s[i] == '#'){
                brk = i;
                break;
            }
            if(s[i]==','){
                len.push_back(stoi(si));
                si = "";
            }
            else    si += s[i];
        }
        int k = brk + 1;
        for(int i=0; i<len.size(); i++){
            string tmp = "";
            int j = 0;
            while(j<len[i]){
                tmp += s[k];
                k++;
                j++;
            }
            ans.push_back(tmp);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/minimum-window-with-characters/submission-3.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        n = len(s)
        i, j = 0, 0
        d1 = {}
        for ch in t:
            d1[ch] = d1.get(ch,0) + 1
        pt = len(d1)
        d2 = {}
        p1, ans = 0, float("infinity")
        xi = []
        while i<n:
            d2[s[i]] = d2.get(s[i], 0) + 1
            if s[i] in d1 and d1[s[i]] == d2[s[i]]:
                p1 += 1
            while j<=i and p1 == pt:
                if ans>i-j+1:
                    ans = i-j+1
                    xi = [j,i]            
                d2[s[j]] -= 1
                if s[j] in d1 and d1[s[j]] > d2[s[j]]:
                    p1 -= 1
                j += 1
            i += 1
        # print(ans)
        return s[xi[0]:xi[1]+1] if len(xi)==2 else ""